community/kernel/src/main/java/org/neo4j/kernel/impl/traversal/BidirectionalTraverserIterator.java

```diff
@@ -91,7 +73,26 @@ public Side( MonoDirectionalTraversalDescription description )
                 new AsOneStartBranch( this, endNodes, end.initialState, start.uniqueness ), end.expander );
 
         this.selector = sideSelector.create( startSelector, endSelector, maxDepth );
-        this.collisionDetector = collisionPolicy.create( collisionEvaluator );
+        this.collisionDetector = buildCollisionDetector( collisionPolicy, collisionEvaluator );
+    }
+
+    private BranchCollisionDetector buildCollisionDetector(
+            org.neo4j.graphdb.traversal.BranchCollisionPolicy collisionPolicy,
+            PathEvaluator collisionEvaluator )
+    {
+        BranchCollisionDetector branchCollisionDetector = collisionPolicy.create( collisionEvaluator );
+        if ( branchCollisionDetector instanceof Filterable )
```

Comment:
As much as `Filterable` is neat, setters makes things slightly harder to reason about than dependency injection. I'd rather we go for a `Predicate` directly into the branch collision detector to avoid this `instanceof` check and also the illegal state check in the branch collision detector implementation.